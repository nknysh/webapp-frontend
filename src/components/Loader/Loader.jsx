import React from 'react';

import peLogo from 'public/img/PE_logo.png';

import { propTypes, defaultProps } from './Loader.props';
import { StyledLoader, LoaderImage, LoaderText } from './Loader.styles';

const Loader = ({ isLoading, text, children }) => {
  if (!isLoading) return children;

  return (
    <StyledLoader>
      <LoaderImage src={peLogo} />
      <LoaderText>{text || 'Loading...'}</LoaderText>
    </StyledLoader>
  );
};

Loader.propTypes = propTypes;
Loader.defaultProps = defaultProps;

export default Loader;
