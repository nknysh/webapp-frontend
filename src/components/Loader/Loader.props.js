import PropTypes from 'prop-types';

export const propTypes = {
  isLoading: PropTypes.bool,
  showSpinner: PropTypes.bool,
  text: PropTypes.string,
  children: PropTypes.oneOfType([PropTypes.element, PropTypes.node]),
};

export const defaultProps = {
  isLoading: false,
  children: null,
  showSpinner: true,
  text: undefined,
};
