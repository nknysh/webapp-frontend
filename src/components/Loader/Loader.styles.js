import styled, { keyframes } from 'styled-components';

import theme from 'styles/theme';
import { Image } from 'styles/elements';
import { P } from 'styles/typography';

const rotate = keyframes`
  from {
    transform: rotate(0deg);
  }

  to {
    transform: rotate(360deg);
  }
`;

export const StyledLoader = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
  width: 100%;
  flex-direction: column;
`;

export const LoaderImage = styled(Image)`
  animation: ${rotate} 1.5s linear infinite;
  transition: ease-in-out 0.25s all;
  display: block;
`;

export const LoaderText = styled(P)`
  font-weight: bold;
  color: ${theme.primary};
  display: block;
`;
