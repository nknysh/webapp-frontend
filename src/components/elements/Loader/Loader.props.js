import PropTypes from 'prop-types';

export const propTypes = {
  isLoading: PropTypes.bool,
  showSpinner: PropTypes.bool,
  text: PropTypes.string,
  children: PropTypes.oneOfType([PropTypes.element, PropTypes.node]),
};

export const defaultProps = {
  isLoading: true,
  children: null,
  showSpinner: true,
  text: undefined,
};
