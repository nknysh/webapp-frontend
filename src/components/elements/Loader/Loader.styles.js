import styled, { keyframes } from 'styled-components';

import theme from 'styles/theme';
import { Image } from 'styles/elements';
import { P } from 'styles/typography';

const rotate = keyframes`
  from {
    transform: rotate(0deg);
  }

  to {
    transform: rotate(360deg);
  }
`;

export const StyledLoader = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
  width: 100%;
  flex-direction: column;
  flex: 1;
`;

export const LoaderImage = styled(Image)`
  animation: ${rotate} 1.5s linear infinite;
  transition: ${theme.defaultTransition};
  display: block;
`;

export const LoaderText = styled(P)`
  font-weight: ${theme.fonts.bold};
  color: ${theme.primary};
  display: block;
`;
