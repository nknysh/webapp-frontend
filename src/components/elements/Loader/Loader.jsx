import React, { useState, Fragment } from 'react';
import { useTranslation } from 'react-i18next';

import { useEffectBoundary } from 'effects';

import peLogo from 'public/assets/img/PE_logo.png';

import { propTypes, defaultProps } from './Loader.props';
import { StyledLoader, LoaderImage, LoaderText } from './Loader.styles';

const Loader = ({ isLoading, text, showSpinner, children, showPrev }) => {
  const { t } = useTranslation();

  const [prevChildren, setPrevChildren] = useState(children);

  useEffectBoundary(() => {
    setPrevChildren(children);
  }, [isLoading]);

  if (!isLoading) {
    !prevChildren && setPrevChildren(children);
    return children;
  }

  return (
    <Fragment>
      {showPrev && prevChildren}
      <StyledLoader data-prev={showPrev}>
        {showSpinner && <LoaderImage src={peLogo} alt={t('title')} />}
        <LoaderText>{text || t('messages.loading')}</LoaderText>
      </StyledLoader>
    </Fragment>
  );
};

Loader.propTypes = propTypes;
Loader.defaultProps = defaultProps;

export default Loader;
