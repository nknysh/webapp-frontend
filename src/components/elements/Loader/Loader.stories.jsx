import React from 'react';

import { storiesOf } from '@storybook/react';
import { withKnobs, text, boolean } from '@storybook/addon-knobs';

import Loader from './Loader';

storiesOf('Loader', module)
  .addDecorator(withKnobs)
  .add('default', () => {
    return (
      <Loader text={text('Loading text')} isLoading={boolean('Is Loading', true)}>
        {text('Content to load', 'Load this content')}
      </Loader>
    );
  });
