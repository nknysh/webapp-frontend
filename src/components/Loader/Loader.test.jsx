import React from 'react';

import Loader from './Loader';

const getComponent = props => shallow(<Loader {...props} />);

describe('<Loader />', () => {
  describe('render', () => {
    it('matches snapshots', () => {
      expect(getComponent()).toMatchSnapshot();
      expect(getComponent({ isLoading: true, text: 'Loading' })).toMatchSnapshot();
      expect(getComponent({ isLoading: true })).toMatchSnapshot();
      expect(getComponent({ isLoading: false, children: 'Loaded content' })).toMatchSnapshot();
    });
  });
});
