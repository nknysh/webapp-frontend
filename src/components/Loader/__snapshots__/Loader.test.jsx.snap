// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Loader /> render matches snapshots 1`] = `""`;

exports[`<Loader /> render matches snapshots 2`] = `
<styled.div>
  <Styled(styled.img)
    src=""
  />
  <Styled(styled.p)>
    Loading
  </Styled(styled.p)>
</styled.div>
`;

exports[`<Loader /> render matches snapshots 3`] = `
<styled.div>
  <Styled(styled.img)
    src=""
  />
  <Styled(styled.p)>
    Loading...
  </Styled(styled.p)>
</styled.div>
`;

exports[`<Loader /> render matches snapshots 4`] = `"Loaded content"`;
